### ISSUE TITLE
Right.Equals returns false for equal anonymous values

### ISSUE THREAD

**Issue by koffmoff** [louthy/language-ext#663]
```cs
var i1 = Right(1);
var i2 = Right(1);
Console.WriteLine(i1.Equals(i2));   // True

var a1 = new { x = 1 };
var a2 = new { x = 1 };
Console.WriteLine(a1.Equals(a2));   // True

var r1 = Right(a1);
var r2 = Right(a2);
Console.WriteLine(r1.Equals(r2));   // False
```

Shouldn't the last comparison be `true` as well?

**Comment by louthy** [comment]
Behind the scenes it uses EqDefault<A> which, if there’s no EqTYPE<A> implementation it falls back to using EqualityComparer<A>.Default not a.Equals(b)

I’m not at a computer right now, so can’t double check

**Comment by koffmoff** [comment]
I get the same result with `EqualityComparer.Default`:

```cs
static void Main()
{
    var i1 = Right(1);
    var i2 = Right(1);
    TestEquals(i1, i2);             // True
    TestEqualityComparer(i1, i2);   // True

    var a1 = new { x = 1 };
    var a2 = new { x = 1 };
    TestEquals(a1, a2);             // True
    TestEqualityComparer(a1, a2);   // True

    var r1 = Right(a1);
    var r2 = Right(a2);
    TestEquals(r1, r2);             // False
    TestEqualityComparer(r1, r2);   // False
}

static void TestEquals<T>(T t1, T t2) =>
    Console.WriteLine(t1.Equals(t2));

static void TestEqualityComparer<T>(T t1, T t2) =>
    Console.WriteLine(EqualityComparer<T>.Default.Equals(t1, t2));

```

**Comment by bender2k14** [comment]
I think the problem is this line
https://github.com/louthy/language-ext/blob/d078300af9ff6db4967a92aa1c591941cd41804f/LanguageExt.Core/ClassInstances/Eq/EqDefault.cs#L25-L27

The following lines are equivalent...
```C#
a1.GetType().GetTypeInfo().ToString()
typeof(A).GetTypeInfo().ToString()
"<>f__AnonymousType0`1[System.Int32]"
```

...for the value of `A` in question.  This causes `isFunc` to be `true` and leads to a reference equality check being performed for `a1` and `a2`.

If I hardcode `isFunc` to `false`, then all of @koffmoff's tests pass.  In fact, all of the tests in `master` continue to pass.  I tried to write some tests that fail wehn `isFunc` is set to `false`, but I was unable to do so.

This idea of `isFunc` was introduced in [`2757a97e`](https://github.com/louthy/language-ext/commit/2757a97eb20ec60e9e23a73268e1a76c49ea2e9b#diff-7cf764fa33aebbc120bb5bd424441e16), expanded in [`c89b3e60`](https://github.com/louthy/language-ext/commit/c89b3e601f904afbd5069da0c1d535823526f291#diff-7cf764fa33aebbc120bb5bd424441e16), and refactored into its current form in [`e949936e`](https://github.com/louthy/language-ext/commit/e949936ee887907f9796d0162ca929477036b975#diff-7cf764fa33aebbc120bb5bd424441e16).

**Comment by bender2k14** [comment]
> I get the same result with `EqualityComparer.Default`.

Wow.  That seems like a bug in .NET to me.

**Comment by bender2k14** [comment]
> I get the same result with `EqualityComparer.Default`:
```C#
...
TestEqualityComparer(r1, r2);   // False
...
```

Are you sure about that?  This tests passes for me.

```C#
[Fact]
public void Test()
{
    var i1 = Right(1);
    var i2 = Right(1);
    DefaultEqualityComparer_Equals(i1, i2);

    var a1 = new { x = 1 };
    var a2 = new { x = 1 };
    DefaultEqualityComparer_Equals(a1, a2);

    var r1 = Right(a1);
    var r2 = Right(a2);
    DefaultEqualityComparer_Equals(r1, r2);
}

static void DefaultEqualityComparer_Equals<T>(T t1, T t2) =>
    Assert.True(EqualityComparer<T>.Default.Equals(t1, t2));
```

**Comment by koffmoff** [comment]
Yes. LanguageExt.Core 3.3.39. Tried both on .NET Core 3.0 and FW 4.7.2. The test fails. Repro: https://github.com/koffmoff/language-ext-equals

**Comment by bender2k14** [comment]
You are right.  That test does _not_ pass for me.  Somehow I got confused and thought that it passed for me.

**Comment by bender2k14** [comment]
And this makes sense, because `EqualityComparer<RightEither<R>>.Default` delegates to...
https://github.com/louthy/language-ext/blob/d078300af9ff6db4967a92aa1c591941cd41804f/LanguageExt.Core/DataTypes/Either-Shared/EitherRight.cs#L182-L183

...which delegates to `EqDefault<R>`, so I think it suffices to consider the analysis I provided in https://github.com/louthy/language-ext/issues/663#issuecomment-554384323.

**Comment by louthy** [comment]
Fixed in `3.3.40`
